#                              ENGAGE 3: Data Challenge

### Submitted by - 
    Pradeep Joshi

### Problem Statement
    1. The file prices.csv describes prices collected for products, represented as UPC, at specific physical store locations, represented as Store ID. The auditors who collected prices at each store are represented as Auditor ID. Store attribute information is described in stores.json, and auditor information is shown in auditors.csv. Can you transform these sources into a cross-tabulation    of regional prices alongside.
    
    2. Do you notice anything that seems off with the data collected?

In [3]:
## Loading Libraries
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import json

In [4]:
## Loading Datasets
prices_df = pd.read_csv('/Users/pradeep/Desktop/Interviews/Engage3/prices.csv')
auditors_df = pd.read_csv('/Users/pradeep/Desktop/Interviews/Engage3/auditors.csv',dtype=object)

with open('/Users/pradeep/Desktop/Interviews/Engage3/stores.json') as f:
    stores = json.load(f)
stores_df = pd.DataFrame(stores)

In [5]:
#prices_df.shape # 12,315 5
#auditors_df.shape # 8 4
#stores_df.shape #28 3
prices_df.head(5)

Auditor ID      Date  Price  Store ID        UPC
0         234  10/18/17  24.95     66999  268588472
1         234  10/27/17  49.71     66999  475245085
2         234  10/20/17  25.75     66999  126967843
3         234  10/23/17  18.81     66999  708930835
4         234  10/23/17  33.32     66999  325885139

In [6]:
auditors_df.head(5)

Auditor ID  First     Last               Region
0        234    Sue    Smith  Northern California
1        536    Bob    Smith  Northern California
2         98   Jack    Smith             New York
3        203   Jill    Smith             New York
4        304  Jerry  Johnson                Texas

In [7]:
stores_df.head(5)

Store ID       Banner               Region
0     66999      Walmart  Northern California
1      4698  Trader Joes  Northern California
2     39482      Safeway  Northern California
3     34957  Whole Foods  Northern California
4     12837      Walmart             New York

In [8]:
prices_df.dtypes

Auditor ID      int64
Date           object
Price         float64
Store ID        int64
UPC             int64
dtype: object

In [9]:
auditors_df.dtypes

Auditor ID    object
First         object
Last          object
Region        object
dtype: object

In [10]:
stores_df.dtypes

Store ID     int64
Banner      object
Region      object
dtype: object

In [11]:
## Changing datatype 
auditors_df['Auditor ID']=auditors_df['Auditor ID'].astype(int)

## merge store and price data by the store ID 
auditors_price_df = pd.merge(auditors_df, prices_df, on='Auditor ID')

## merge previous data with auditor data by auditor ID
auditors_price_store_df = pd.merge(auditors_price_df, stores_df, on='Store ID')

auditors_price_store_df.head()

Auditor ID First   Last             Region_x      Date  Price  Store ID  \
0         234   Sue  Smith  Northern California  10/18/17  24.95     66999   
1         234   Sue  Smith  Northern California  10/27/17  49.71     66999   
2         234   Sue  Smith  Northern California  10/20/17  25.75     66999   
3         234   Sue  Smith  Northern California  10/23/17  18.81     66999   
4         234   Sue  Smith  Northern California  10/23/17  33.32     66999   

         UPC   Banner             Region_y  
0  268588472  Walmart  Northern California  
1  475245085  Walmart  Northern California  
2  126967843  Walmart  Northern California  
3  708930835  Walmart  Northern California  
4  325885139  Walmart  Northern California

In [12]:
auditors_price_store_df

Auditor ID First     Last             Region_x      Date  Price  \
0             234   Sue    Smith  Northern California  10/18/17  24.95   
1             234   Sue    Smith  Northern California  10/27/17  49.71   
2             234   Sue    Smith  Northern California  10/20/17  25.75   
3             234   Sue    Smith  Northern California  10/23/17  18.81   
4             234   Sue    Smith  Northern California  10/23/17  33.32   
...           ...   ...      ...                  ...       ...    ...   
11506         713  Dave  Johnson               Kansas  10/19/17   1.99   
11507         713  Dave  Johnson               Kansas  10/29/17   1.99   
11508         713  Dave  Johnson               Kansas  10/25/17   1.99   
11509         713  Dave  Johnson               Kansas  10/21/17   1.99   
11510         713  Dave  Johnson               Kansas  10/21/17   1.99   

       Store ID        UPC       Banner             Region_y  
0         66999  268588472      Walmart  Northern California  
1         66999  475245085      Walmart  Northern California  
2         66999  126967843      Walmart  Northern California  
3         66999  708930835      Walmart  Northern California  
4         66999  325885139      Walmart  Northern California  
...         ...        ...          ...                  ...  
11506     39287  950498192  Whole Foods               Kansas  
11507     39287  303796970  Whole Foods               Kansas  
11508     39287  125829867  Whole Foods               Kansas  
11509     39287  912873332  Whole Foods               Kansas  
11510     39287  163809044  Whole Foods               Kansas  

[11511 rows x 10 columns]

In [13]:
## drop the last col which is aduitor's region, it get less confuse with the price region
#auditors_price_store_df = auditors_price_store_df.drop(['Region_y'], axis=1) 

## change the region_x to the region
auditors_price_store_df = auditors_price_store_df.rename(columns={"Region_x": "Region_Auditor"})
auditors_price_store_df = auditors_price_store_df.rename(columns={"Region_y": "Region_Store"})

auditors_price_store_df.head()

Auditor ID First   Last       Region_Auditor      Date  Price  Store ID  \
0         234   Sue  Smith  Northern California  10/18/17  24.95     66999   
1         234   Sue  Smith  Northern California  10/27/17  49.71     66999   
2         234   Sue  Smith  Northern California  10/20/17  25.75     66999   
3         234   Sue  Smith  Northern California  10/23/17  18.81     66999   
4         234   Sue  Smith  Northern California  10/23/17  33.32     66999   

         UPC   Banner         Region_Store  
0  268588472  Walmart  Northern California  
1  475245085  Walmart  Northern California  
2  126967843  Walmart  Northern California  
3  708930835  Walmart  Northern California  
4  325885139  Walmart  Northern California

In [14]:
## Unique UPC
len(auditors_price_store_df['UPC'].unique())

1000

In [15]:
## Unique Stores
len(auditors_price_store_df['Store ID'].unique())

18

In [16]:
## Unique Regions
auditors_price_store_df['Region_Auditor'].unique()

array(['Northern California', 'New York', 'Texas', 'Kansas'], dtype=object)

In [17]:
## Where Auditor and Store Regions are Different
region_Diff = auditors_price_store_df[auditors_price_store_df['Region_Auditor'] != auditors_price_store_df['Region_Store']]
region_Diff

Auditor ID First   Last Region_Auditor      Date  Price  Store ID  \
1328          98  Jack  Smith       New York  10/25/17  16.89     39482   
1329          98  Jack  Smith       New York  10/20/17   8.89     39482   
1330          98  Jack  Smith       New York  10/22/17  57.69     39482   
1331          98  Jack  Smith       New York  10/17/17   3.49     39482   
1332          98  Jack  Smith       New York  10/21/17   9.39     39482   
...          ...   ...    ...            ...       ...    ...       ...   
1629          98  Jack  Smith       New York  10/27/17   9.19     39482   
1630          98  Jack  Smith       New York  10/18/17  45.69     39482   
1631          98  Jack  Smith       New York  10/23/17  15.19     39482   
1632          98  Jack  Smith       New York  10/24/17  17.49     39482   
1633          98  Jack  Smith       New York  10/16/17  36.59     39482   

            UPC   Banner         Region_Store  
1328  286906735  Safeway  Northern California  
1329  672429356  Safeway  Northern California  
1330  506754007  Safeway  Northern California  
1331  421047592  Safeway  Northern California  
1332  555173446  Safeway  Northern California  
...         ...      ...                  ...  
1629  748626363  Safeway  Northern California  
1630  872889619  Safeway  Northern California  
1631  581030587  Safeway  Northern California  
1632  790024227  Safeway  Northern California  
1633  783399454  Safeway  Northern California  

[306 rows x 10 columns]

In [18]:
## 39482 is the ID when regions are different
region_Diff['Store ID'].unique()

array([39482])

In [19]:
## Analyzing 39482
region_Diff.loc[region_Diff['Store ID']==39482]

Auditor ID First   Last Region_Auditor      Date  Price  Store ID  \
1328          98  Jack  Smith       New York  10/25/17  16.89     39482   
1329          98  Jack  Smith       New York  10/20/17   8.89     39482   
1330          98  Jack  Smith       New York  10/22/17  57.69     39482   
1331          98  Jack  Smith       New York  10/17/17   3.49     39482   
1332          98  Jack  Smith       New York  10/21/17   9.39     39482   
...          ...   ...    ...            ...       ...    ...       ...   
1629          98  Jack  Smith       New York  10/27/17   9.19     39482   
1630          98  Jack  Smith       New York  10/18/17  45.69     39482   
1631          98  Jack  Smith       New York  10/23/17  15.19     39482   
1632          98  Jack  Smith       New York  10/24/17  17.49     39482   
1633          98  Jack  Smith       New York  10/16/17  36.59     39482   

            UPC   Banner         Region_Store  
1328  286906735  Safeway  Northern California  
1329  672429356  Safeway  Northern California  
1330  506754007  Safeway  Northern California  
1331  421047592  Safeway  Northern California  
1332  555173446  Safeway  Northern California  
...         ...      ...                  ...  
1629  748626363  Safeway  Northern California  
1630  872889619  Safeway  Northern California  
1631  581030587  Safeway  Northern California  
1632  790024227  Safeway  Northern California  
1633  783399454  Safeway  Northern California  

[306 rows x 10 columns]

In [20]:
region_Diff.loc[region_Diff['Store ID']==39482]['Auditor ID'].unique()

array([98])

In [21]:
stores_df[stores_df['Store ID']==39482]

Store ID   Banner               Region
2     39482  Safeway  Northern California

In [22]:
auditors_df[auditors_df['Auditor ID']==98]

Auditor ID First   Last    Region
2          98  Jack  Smith  New York

In [23]:
## Checking rows which don't have any information in the table
region_Diff.loc[region_Diff['Store ID']!=39482]

Empty DataFrame
Columns: [Auditor ID, First, Last, Region_Auditor, Date, Price, Store ID, UPC, Banner, Region_Store]
Index: []

In [24]:
## Croos-tabulation of Regional Prices
crosstab_df = pd.crosstab(index=[auditors_price_store_df['Banner'],auditors_price_store_df['UPC']], \
                         columns=auditors_price_store_df['Region_Store'], values=auditors_price_store_df['Price'],\
                         aggfunc='median').fillna('-')
crosstab_df

Region_Store          Kansas New York Northern California  Texas
Banner      UPC                                                 
Safeway     11873171       -        -                6.09   5.19
            15052612   53.99        -                   -  54.49
            16482322   17.89        -                   -  18.09
            16729338    7.99        -                9.39   8.09
            16829288    3.59        -                4.19   3.59
...                      ...      ...                 ...    ...
Whole Foods 995798889   1.99    62.39               70.39  60.59
            996262978   1.99    14.39               16.19  13.89
            996849471   1.99    12.79                   -      -
            998831540   1.99    39.99                   -  38.79
            999185078   1.99    58.09               65.49      -

[4931 rows x 4 columns]

In [25]:
crosstab_df.describe()

Region_Store Kansas New York Northern California Texas
count          4931     4931                4931  4931
unique         1103     1395                 984  1157
top               -        -                   -     -
freq           1844     1906                3297  1166

In [26]:
# save to the csv file
crosstab_df.to_csv("Engage3_CrossTab.csv", index=False)

## CONCLUSION    
    From the cross tab it's evident that in Kansas all the products at Whole Foods have the same price of 1.99. This seems to be an anomaly or some sort of input error